### PR TITLE
listener: add optional rate and cleanup existing arguments

### PR DESCRIPTION
**Pull request by dagar**
This adds an optional rate (-r) argument to listener command and updates the existing awkward arguments for instance and number of messages.

image

@CarlOlsson I believe you asked for this a long time ago.

### CODE REVIEW

**Comment by mhkabir** [comment]
Nice, can you also add an option to just wait for messages on a (potentially yet to be published) topic and keep printing them once received, till Ctrl-C?

**Comment by dagar** [comment]
How about something like `-w` (for wait) combined with the existing options?

**Comment by dagar** [comment]
That could be the default if `-r` is specified without `-n`. The only reason I gave it an arbitrary limit was that ctrl-c in pxh (px4 shell?) needs work.

**Comment by CarlOlsson** [comment]
> @CarlOlsson I believe you asked for this a long time ago.

Yes really nice!

File: src/systemcmds/topic_listener/generate_listener.py [hunk 1]
```diff
@@ -124,35 +123,42 @@
...
-print("\tunsigned topic_instance = (argc > 3) ? atoi(argv[3]) : 0;\n")
+
 for index, (m, t) in enumerate(zip(messages, topics)):
 	if index == 0:
-		print("\tif (strncmp(argv[1],\"%s\",50) == 0) {" % t)
+		print("\tif (strncmp(topic_name,\"%s\", %d) == 0) {" % (t, len(t)))
 	else:
-		print("\t} else if (strncmp(argv[1],\"%s\",50) == 0) {" % t)
-	print("\t\tlisten_%s(num_msgs, topic_instance);" % t)
+		print("\t} else if (strncmp(topic_name,\"%s\", %d) == 0) {" % (t, len(t)))
```
Comment (bkueng) [inline src/systemcmds/topic_listener/generate_listener.py]: Can you use `strcmp`? We know that both are null-terminated strings and using `strncmp` can lead to unwanted prefix matches.

**Comment by dagar** [comment]
@mhkabir we need to look at getting ctrl-c (SIGINT signal handler) working for your suggestions.

**Comment by bkueng** [comment]
@dagar can you look into the CI failure? This is good to go other than that.

**Comment by dagar** [comment]
Yes I saw it earlier, but my first attempt at getting the uorb graph generator to ignore listener failed. I'll take another look at it later today.

**Pull request merged**